# Comparison of dataset generation parameters
This notebook compares two different backbones, qiskit and qudit-kit, for dataset generation. Moreover, it evaluates the impact of (not) optimizing circuits using `qiskit.transpile` during generation.

Comparison is done regarding
1. Time for generation and
2. Number of unqiue circuits vs generated circuits.

In [1]:
import sys
sys.path.append("..")

import time
import logging
logging.getLogger("stevedore").setLevel("WARNING")

from scripts.generate_dataset import main

In [3]:
from hydra import initialize, compose

with initialize(version_base=None, config_path="../conf"):
    cfg = compose(config_name="config.yaml")

# SRV task

## Results

| Backbone | Optimized | Num Samples | Time | Percentage of unique samples
| --- | --- | --- | --- | --- |
| Qiskit | True | 10k | 50.17s | 98.34%
| Qiskit | True | 100k | 511.54s | 94.47%
| Qiskit | False | 10k | 32.10s | 99.95%
| Qiskit | False | 100k | 710.96s | 99.75%
| --- | --- | --- | --- | --- |
| Qudit-Kit | True | 10k | 130.67s | 99.93%
| Qudit-Kit | True | 100k | 1182.70s | 99.74%
| Qudit-Kit | False | 10k | 31.94s | 99.98%
| Qudit-Kit | False | 100k | 298.21s | 99.75%

## Import default configs

- gate_set: H, CX
- num_qubits: 5
- num_samples: 10000
- min_gates: 4
- max_gates: 20
- backbone: qiskit
- condition_type: SRV
- optimized: True
- output_path: ./datasets/srv_dataset
- device: auto

## Backbone: Qiskit

### Qiskit with optimization, 10.000 samples

#### Results
- Time: 50.17 seconds
- generated qubits: 9978
- unique qubits: 9812 (98.34%)

In [3]:
cfg.datasets.backbone = "qiskit"
cfg.datasets.optimized = True
cfg.datasets.num_samples = 10000
cfg.datasets.output_path = "./tests/datasets/qiskit_optimized_10k"

main(cfg)

2025-12-07 21:44:28 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx'], 'num_qubits': 5, 'num_samples': 10000, 'min_gates': 4, 'max_gates': 20, 'backbone': 'qiskit', 'condition_type': 'SRV', 'optimized': True, 'output_path': './tests/datasets/qiskit_optimized_10k', 'device': 'auto'}
2025-12-07 21:44:28 - quantum_diffusion.data.dataset - INFO - Generating dataset with 10000 samples, 5 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2025-12-07 21:44:28 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for SRV...
[INFO]: Generated 9973 valid circuits.
[INFO]: After filtering unique circuits: 9802.
[INFO]: Saving tensor to `tests/datasets/qiskit_optimized_10k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_optimized_10k/dataset/ds_y.safetensors`.
2025-12-07 21:46:32 -

### Qiskit with optimization, 100.000 samples

#### Results
- Time: 511.54 seconds
- generated qubits: 99846
- unique qubits: 94328 (94.47%)

In [4]:
cfg.datasets.backbone = "qiskit"
cfg.datasets.optimized = True
cfg.datasets.num_samples: 100000
cfg.datasets.output_path = "./tests/datasets/qiskit_optimized_100k"

main(cfg)

2025-12-07 21:46:32 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx'], 'num_qubits': 5, 'num_samples': 10000, 'min_gates': 4, 'max_gates': 20, 'backbone': 'qiskit', 'condition_type': 'SRV', 'optimized': True, 'output_path': './tests/datasets/qiskit_optimized_100k', 'device': 'auto'}
2025-12-07 21:46:32 - quantum_diffusion.data.dataset - INFO - Generating dataset with 10000 samples, 5 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2025-12-07 21:46:32 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for SRV...
[INFO]: Generated 9977 valid circuits.
[INFO]: After filtering unique circuits: 9763.
[INFO]: Saving tensor to `tests/datasets/qiskit_optimized_100k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_optimized_100k/dataset/ds_y.safetensors`.
2025-12-07 21:48:0

### Qiskit without optimization, 10.000 samples

#### Results
- Time: 32.10 seconds
- generated qubits: 9984
- unique qubits: 9979 (99.95%)

In [5]:
cfg.datasets.backbone = "qiskit"
cfg.datasets.optimized = False
cfg.datasets.num_samples = 10000
cfg.datasets.output_path = "./tests/datasets/qiskit_not_optimized_10k"

main(cfg)

2025-12-07 21:48:07 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx'], 'num_qubits': 5, 'num_samples': 10000, 'min_gates': 4, 'max_gates': 20, 'backbone': 'qiskit', 'condition_type': 'SRV', 'optimized': False, 'output_path': './tests/datasets/qiskit_not_optimized_10k', 'device': 'auto'}
2025-12-07 21:48:07 - quantum_diffusion.data.dataset - INFO - Generating dataset with 10000 samples, 5 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2025-12-07 21:48:07 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for SRV...
[INFO]: Generated 9984 valid circuits.
[INFO]: After filtering unique circuits: 9982.
[INFO]: Saving tensor to `tests/datasets/qiskit_not_optimized_10k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_not_optimized_10k/dataset/ds_y.safetensors`.
2025-12-

### Qiskit without optimization, 100.000 samples

#### Results
- Time: 710.96 seconds
- generated qubits: 99968
- unique qubits: 99715 (99.75%)

In [6]:
cfg.datasets.backbone = "qiskit"
cfg.datasets.optimized = False
cfg.datasets.num_samples = 100000
cfg.datasets.output_path = "./tests/datasets/qiskit_not_optimized_100k"

main(cfg)

2025-12-07 21:49:10 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx'], 'num_qubits': 5, 'num_samples': 100000, 'min_gates': 4, 'max_gates': 20, 'backbone': 'qiskit', 'condition_type': 'SRV', 'optimized': False, 'output_path': './tests/datasets/qiskit_not_optimized_100k', 'device': 'auto'}
2025-12-07 21:49:10 - quantum_diffusion.data.dataset - INFO - Generating dataset with 100000 samples, 5 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2025-12-07 21:49:10 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for SRV...
[INFO]: Generated 99968 valid circuits.
[INFO]: After filtering unique circuits: 99713.
[INFO]: Saving tensor to `tests/datasets/qiskit_not_optimized_100k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_not_optimized_100k/dataset/ds_y.safetensors`.
2

## Backbone: Quditkit

### Quditkit with optimization, 10.000 samples

#### Results
- Time: 130.67 seconds
- generated qubits: 9984
- unique qubits: 9977 (99.93%)

In [7]:
cfg.datasets.backbone = "quditkit"
cfg.datasets.optimized = True
cfg.datasets.num_samples = 10000
cfg.datasets.output_path = "./tests/datasets/quditkit_optimized_10k"

main(cfg)

2025-12-07 22:00:09 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx'], 'num_qubits': 5, 'num_samples': 10000, 'min_gates': 4, 'max_gates': 20, 'backbone': 'quditkit', 'condition_type': 'SRV', 'optimized': True, 'output_path': './tests/datasets/quditkit_optimized_10k', 'device': 'auto'}
2025-12-07 22:00:09 - quantum_diffusion.data.dataset - INFO - Generating dataset with 10000 samples, 5 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2025-12-07 22:00:09 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for SRV...
[INFO]: Generated 9984 valid circuits.
[INFO]: After filtering unique circuits: 9980.
[INFO]: Saving tensor to `tests/datasets/quditkit_optimized_10k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_optimized_10k/dataset/ds_y.safetensors`.
2025-12-07 22

### Quditkit with optimization, 100.000 samples

#### Results
- Time: 1182.70 seconds
- generated qubits: 99967
- unique qubits: 99706 (99.74 %)

In [8]:
cfg.datasets.backbone = "quditkit"
cfg.datasets.optimized = True
cfg.datasets.num_samples = 100000
cfg.datasets.output_path = "./tests/datasets/quditkit_optimized_100k"

main(cfg)

2025-12-07 22:02:09 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx'], 'num_qubits': 5, 'num_samples': 100000, 'min_gates': 4, 'max_gates': 20, 'backbone': 'quditkit', 'condition_type': 'SRV', 'optimized': True, 'output_path': './tests/datasets/quditkit_optimized_100k', 'device': 'auto'}
2025-12-07 22:02:09 - quantum_diffusion.data.dataset - INFO - Generating dataset with 100000 samples, 5 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2025-12-07 22:02:09 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for SRV...
[INFO]: Generated 99968 valid circuits.
[INFO]: After filtering unique circuits: 99732.
[INFO]: Saving tensor to `tests/datasets/quditkit_optimized_100k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_optimized_100k/dataset/ds_y.safetensors`.
2025-1

### Quditkit without optimization, 10.000 samples

#### Results
- Time: 31.94 seconds
- generated qubits: 9983
- unique qubits: 9981 (99.98 %)

In [9]:
cfg.datasets.backbone = "quditkit"
cfg.datasets.optimized = False
cfg.datasets.num_samples = 10000
cfg.datasets.output_path = "./tests/datasets/quditkit_not_optimized_10k"

main(cfg)

2025-12-07 22:19:25 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx'], 'num_qubits': 5, 'num_samples': 10000, 'min_gates': 4, 'max_gates': 20, 'backbone': 'quditkit', 'condition_type': 'SRV', 'optimized': False, 'output_path': './tests/datasets/quditkit_not_optimized_10k', 'device': 'auto'}
2025-12-07 22:19:25 - quantum_diffusion.data.dataset - INFO - Generating dataset with 10000 samples, 5 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2025-12-07 22:19:25 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for SRV...
[INFO]: Generated 9984 valid circuits.
[INFO]: After filtering unique circuits: 9981.
[INFO]: Saving tensor to `tests/datasets/quditkit_not_optimized_10k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_not_optimized_10k/dataset/ds_y.safetensors`.


### Quditkit without optimization, 100.000 samples

#### Results
- Time: 298.21 seconds
- generated qubits: 99968 
- unique qubits: 99714 (99.75 %)

In [10]:
cfg.datasets.backbone = "quditkit"
cfg.datasets.optimized = False
cfg.datasets.num_samples = 100000
cfg.datasets.output_path = "./tests/datasets/quditkit_not_optimized_100k"

main(cfg)

2025-12-07 22:20:09 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx'], 'num_qubits': 5, 'num_samples': 100000, 'min_gates': 4, 'max_gates': 20, 'backbone': 'quditkit', 'condition_type': 'SRV', 'optimized': False, 'output_path': './tests/datasets/quditkit_not_optimized_100k', 'device': 'auto'}
2025-12-07 22:20:09 - quantum_diffusion.data.dataset - INFO - Generating dataset with 100000 samples, 5 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2025-12-07 22:20:09 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for SRV...
[INFO]: Generated 99967 valid circuits.
[INFO]: After filtering unique circuits: 99706.
[INFO]: Saving tensor to `tests/datasets/quditkit_not_optimized_100k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_not_optimized_100k/dataset/ds_y.safeten

---

# Unitary compilation task

CAVE: The times are significantly lower then for the SRV task, since here only 3 qubits were used.

## Results

| Backbone | Optimized | Num Samples | Time | Percentage of unique samples
| --- | --- | --- | --- | --- |
| Qiskit | True | 10k | 50.17s | 98.34%
| Qiskit | True | 100k | 511.54s | 94.47%
| Qiskit | False | 10k | 32.10s | 99.95%
| Qiskit | False | 100k | 710.96s | 99.75%
| --- | --- | --- | --- | --- |
| Qudit-Kit | True | 10k | 130.67s | 99.93%
| Qudit-Kit | True | 100k | 1182.70s | 99.74%
| Qudit-Kit | False | 10k | 31.94s | 99.98%
| Qudit-Kit | False | 100k | 298.21s | 99.75%

## Import unitary configs

- qiskit gate_set: H, CX, Z, X, CCX, SWAP
- quditkit gate_set: H, CX, Z, X, SWAP (`ccx` omitted)
- num_qubits: 3
- num_samples: 10000
- min_gates: 2
- max_gates: 12
- condition_type: UNITARY
- backbone: qiskit
- optimized: True
- output_path: ./datasets/paper_qiskit/unitary_3q_dataset
- device: auto

In [4]:
cfg.datasets.gate_set = ["h", "cx", "z", "x", "ccx", "swap"]
cfg.datasets.num_qubits = 3
cfg.datasets.min_gates = 2
cfg.datasets.max_gates = 12
cfg.datasets.condition_type = "UNITARY"
cfg.datasets.backbone = "qiskit"
cfg.datasets.optimized = True
cfg.datasets.num_samples = 10000
cfg.datasets.output_path = "./tests/datasets/qiskit_unitary_optimized_10k"

## Backbone: Qiskit

### Qiskit with optimization, 10.000 samples

#### Results
- Time: 19.21 seconds
- generated circuits: 9833
- unique circuits: 8603 (87.49%)

In [5]:
cfg.datasets.gate_set = ["h", "cx", "z", "x", "ccx", "swap"]
cfg.datasets.condition_type = "UNITARY"
cfg.datasets.backbone = "qiskit"
cfg.datasets.optimized = True
cfg.datasets.num_samples = 10000
cfg.datasets.output_path = "./tests/datasets/qiskit_unitary_optimized_10k"

main(cfg)

[INFO]: Cuda device has a capability of 8.6 (>= 8), allowing tf32 matmul.
2026-03-08 18:24:06 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx', 'z', 'x', 'ccx', 'swap'], 'num_qubits': 3, 'num_samples': 10000, 'min_gates': 2, 'max_gates': 12, 'backbone': 'qiskit', 'condition_type': 'UNITARY', 'optimized': True, 'output_path': './tests/datasets/qiskit_unitary_optimized_10k', 'device': 'auto', 'n_jobs': 4}
2026-03-08 18:24:06 - quantum_diffusion.data.dataset - INFO - Generating dataset with 10000 samples, 3 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2026-03-08 18:24:07 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for UNITARY...
[INFO]: Generated 9833 valid circuits.
[INFO]: After filtering unique circuits: 8603.


  0%|          | 0/1 [00:00<?, ?it/s]

 - balance_tensor_dataset, njobs=1, number of samples=8603
 - uniquify_tensor_dataset, number of samples now 8603
 - balancing


  0%|          | 0/57 [00:00<?, ?it/s]

 - dataset size after balancing 8603
[INFO]: allocate memory for U (8603, 2, 8, 8) on cpu approx. 0.004 GB
Split: Train 8173 - Test 430 

[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_optimized_10k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_optimized_10k/dataset/ds_y.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_optimized_10k/dataset/ds_U.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_optimized_10k/dataset/ds_z.safetensors`.
2026-03-08 18:24:26 - quantum_diffusion.data.dataset - INFO - UNITARY dataset saved to tests/datasets/qiskit_unitary_optimized_10k
2026-03-08 18:24:26 - scripts.generate_dataset - INFO - Dataset generation completed successfully! Took 19.21 seconds.
2026-03-08 18:24:26 - scripts.generate_dataset - INFO - UNITARY: saved to tests/datasets/qiskit_unitary_optimized_10k

DATASET GENERATION SUMMARY
Total time elapsed: 19.21 seconds
Output paths:
  UNITARY: tests/datasets/qisk

### Qiskit with optimization, 100.000 samples

#### Results
- Time: 99.10 seconds
- generated circuits: 98641
- unique circuits: 75856 (76.90%)

In [6]:
cfg.datasets.gate_set = ["h", "cx", "z", "x", "ccx", "swap"]
cfg.datasets.condition_type = "UNITARY"
cfg.datasets.backbone = "qiskit"
cfg.datasets.optimized = True
cfg.datasets.num_samples = 100000
cfg.datasets.output_path = "./tests/datasets/qiskit_unitary_optimized_100k"

main(cfg)

[INFO]: Cuda device has a capability of 8.6 (>= 8), allowing tf32 matmul.
2026-03-08 18:24:26 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx', 'z', 'x', 'ccx', 'swap'], 'num_qubits': 3, 'num_samples': 100000, 'min_gates': 2, 'max_gates': 12, 'backbone': 'qiskit', 'condition_type': 'UNITARY', 'optimized': True, 'output_path': './tests/datasets/qiskit_unitary_optimized_100k', 'device': 'auto', 'n_jobs': 4}
2026-03-08 18:24:26 - quantum_diffusion.data.dataset - INFO - Generating dataset with 100000 samples, 3 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2026-03-08 18:24:26 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for UNITARY...
[INFO]: Generated 98641 valid circuits.
[INFO]: After filtering unique circuits: 75856.


  0%|          | 0/1 [00:00<?, ?it/s]

 - balance_tensor_dataset, njobs=1, number of samples=75856
 - uniquify_tensor_dataset, number of samples now 75856
 - balancing


  0%|          | 0/57 [00:00<?, ?it/s]

 - dataset size after balancing 75856
[INFO]: allocate memory for U (75856, 2, 8, 8) on cpu approx. 0.036 GB
Split: Train 72064 - Test 3792 

[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_optimized_100k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_optimized_100k/dataset/ds_y.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_optimized_100k/dataset/ds_U.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_optimized_100k/dataset/ds_z.safetensors`.
2026-03-08 18:26:05 - quantum_diffusion.data.dataset - INFO - UNITARY dataset saved to tests/datasets/qiskit_unitary_optimized_100k
2026-03-08 18:26:05 - scripts.generate_dataset - INFO - Dataset generation completed successfully! Took 99.10 seconds.
2026-03-08 18:26:05 - scripts.generate_dataset - INFO - UNITARY: saved to tests/datasets/qiskit_unitary_optimized_100k

DATASET GENERATION SUMMARY
Total time elapsed: 99.10 seconds
Output paths:
  UNITARY: tests/dat

### Qiskit without optimization, 10.000 samples

#### Results
- Time: 4.94 seconds
- generated circuits: 9965
- unique circuits: 9331 (93.64%)

In [7]:
cfg.datasets.gate_set = ["h", "cx", "z", "x", "ccx", "swap"]
cfg.datasets.condition_type = "UNITARY"
cfg.datasets.backbone = "qiskit"
cfg.datasets.optimized = False
cfg.datasets.num_samples = 10000
cfg.datasets.output_path = "./tests/datasets/qiskit_unitary_not_optimized_10k"

main(cfg)

[INFO]: Cuda device has a capability of 8.6 (>= 8), allowing tf32 matmul.
2026-03-08 18:26:05 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx', 'z', 'x', 'ccx', 'swap'], 'num_qubits': 3, 'num_samples': 10000, 'min_gates': 2, 'max_gates': 12, 'backbone': 'qiskit', 'condition_type': 'UNITARY', 'optimized': False, 'output_path': './tests/datasets/qiskit_unitary_not_optimized_10k', 'device': 'auto', 'n_jobs': 4}
2026-03-08 18:26:05 - quantum_diffusion.data.dataset - INFO - Generating dataset with 10000 samples, 3 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2026-03-08 18:26:05 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for UNITARY...
[INFO]: Generated 9965 valid circuits.
[INFO]: After filtering unique circuits: 9331.


  0%|          | 0/1 [00:00<?, ?it/s]

 - balance_tensor_dataset, njobs=1, number of samples=9331
 - uniquify_tensor_dataset, number of samples now 9331
 - balancing


  0%|          | 0/57 [00:00<?, ?it/s]

 - dataset size after balancing 9331
[INFO]: allocate memory for U (9331, 2, 8, 8) on cpu approx. 0.004 GB
Split: Train 8865 - Test 466 

[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_not_optimized_10k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_not_optimized_10k/dataset/ds_y.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_not_optimized_10k/dataset/ds_U.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_not_optimized_10k/dataset/ds_z.safetensors`.
2026-03-08 18:26:10 - quantum_diffusion.data.dataset - INFO - UNITARY dataset saved to tests/datasets/qiskit_unitary_not_optimized_10k
2026-03-08 18:26:10 - scripts.generate_dataset - INFO - Dataset generation completed successfully! Took 4.94 seconds.
2026-03-08 18:26:10 - scripts.generate_dataset - INFO - UNITARY: saved to tests/datasets/qiskit_unitary_not_optimized_10k

DATASET GENERATION SUMMARY
Total time elapsed: 4.94 seconds
Output paths:
  UNITAR

### Qiskit without optimization, 100.000 samples

#### Results
- Time: 42.37 seconds
- generated circuits: 99793
- unique circuits: 86816 (87.00%)

In [8]:
cfg.datasets.gate_set = ["h", "cx", "z", "x", "ccx", "swap"]
cfg.datasets.condition_type = "UNITARY"
cfg.datasets.backbone = "qiskit"
cfg.datasets.optimized = False
cfg.datasets.num_samples = 100000
cfg.datasets.output_path = "./tests/datasets/qiskit_unitary_not_optimized_100k"

main(cfg)

[INFO]: Cuda device has a capability of 8.6 (>= 8), allowing tf32 matmul.
2026-03-08 18:26:10 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx', 'z', 'x', 'ccx', 'swap'], 'num_qubits': 3, 'num_samples': 100000, 'min_gates': 2, 'max_gates': 12, 'backbone': 'qiskit', 'condition_type': 'UNITARY', 'optimized': False, 'output_path': './tests/datasets/qiskit_unitary_not_optimized_100k', 'device': 'auto', 'n_jobs': 4}
2026-03-08 18:26:10 - quantum_diffusion.data.dataset - INFO - Generating dataset with 100000 samples, 3 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2026-03-08 18:26:10 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for UNITARY...
[INFO]: Generated 99793 valid circuits.
[INFO]: After filtering unique circuits: 86816.


  0%|          | 0/1 [00:00<?, ?it/s]

 - balance_tensor_dataset, njobs=1, number of samples=86816
 - uniquify_tensor_dataset, number of samples now 86816
 - balancing


  0%|          | 0/57 [00:00<?, ?it/s]

 - dataset size after balancing 86816
[INFO]: allocate memory for U (86816, 2, 8, 8) on cpu approx. 0.041 GB
Split: Train 82476 - Test 4340 

[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_not_optimized_100k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_not_optimized_100k/dataset/ds_y.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_not_optimized_100k/dataset/ds_U.safetensors`.
[INFO]: Saving tensor to `tests/datasets/qiskit_unitary_not_optimized_100k/dataset/ds_z.safetensors`.
2026-03-08 18:26:52 - quantum_diffusion.data.dataset - INFO - UNITARY dataset saved to tests/datasets/qiskit_unitary_not_optimized_100k
2026-03-08 18:26:52 - scripts.generate_dataset - INFO - Dataset generation completed successfully! Took 42.37 seconds.
2026-03-08 18:26:52 - scripts.generate_dataset - INFO - UNITARY: saved to tests/datasets/qiskit_unitary_not_optimized_100k

DATASET GENERATION SUMMARY
Total time elapsed: 42.37 seconds
Output pat

## Backbone: Quditkit

### Quditkit with optimization, 10.000 samples

#### Results
- Time: 13.34 seconds
- generated circuits: 9970
- unique circuits: 9373 (94.01%)

In [9]:
cfg.datasets.gate_set = ["h", "cx", "z", "x", "swap"]
cfg.datasets.condition_type = "UNITARY"
cfg.datasets.backbone = "quditkit"
cfg.datasets.optimized = True
cfg.datasets.num_samples = 10000
cfg.datasets.output_path = "./tests/datasets/quditkit_unitary_optimized_10k"

main(cfg)

[INFO]: Cuda device has a capability of 8.6 (>= 8), allowing tf32 matmul.
2026-03-08 18:26:52 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx', 'z', 'x', 'swap'], 'num_qubits': 3, 'num_samples': 10000, 'min_gates': 2, 'max_gates': 12, 'backbone': 'quditkit', 'condition_type': 'UNITARY', 'optimized': True, 'output_path': './tests/datasets/quditkit_unitary_optimized_10k', 'device': 'auto', 'n_jobs': 4}
2026-03-08 18:26:52 - quantum_diffusion.data.dataset - INFO - Generating dataset with 10000 samples, 3 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2026-03-08 18:26:52 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for UNITARY...
[INFO]: Generated 9970 valid circuits.
[INFO]: After filtering unique circuits: 9373.


  0%|          | 0/1 [00:00<?, ?it/s]

 - balance_tensor_dataset, njobs=1, number of samples=9373
 - uniquify_tensor_dataset, number of samples now 9373
 - balancing


  0%|          | 0/26 [00:00<?, ?it/s]

 - dataset size after balancing 9373
[INFO]: allocate memory for U (9373, 2, 8, 8) on cpu approx. 0.004 GB
Split: Train 8905 - Test 468 

[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_optimized_10k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_optimized_10k/dataset/ds_y.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_optimized_10k/dataset/ds_U.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_optimized_10k/dataset/ds_z.safetensors`.
2026-03-08 18:27:05 - quantum_diffusion.data.dataset - INFO - UNITARY dataset saved to tests/datasets/quditkit_unitary_optimized_10k
2026-03-08 18:27:05 - scripts.generate_dataset - INFO - Dataset generation completed successfully! Took 13.34 seconds.
2026-03-08 18:27:05 - scripts.generate_dataset - INFO - UNITARY: saved to tests/datasets/quditkit_unitary_optimized_10k

DATASET GENERATION SUMMARY
Total time elapsed: 13.34 seconds
Output paths:
  UNITARY: tests/d

### Quditkit with optimization, 100.000 samples

#### Results
- Time: 114.19 seconds
- generated circuits: 99792
- unique circuits: 86536 (86.72%)

In [10]:
cfg.datasets.gate_set = ["h", "cx", "z", "x", "swap"]
cfg.datasets.condition_type = "UNITARY"
cfg.datasets.backbone = "quditkit"
cfg.datasets.optimized = True
cfg.datasets.num_samples = 100000
cfg.datasets.output_path = "./tests/datasets/quditkit_unitary_optimized_100k"

main(cfg)

[INFO]: Cuda device has a capability of 8.6 (>= 8), allowing tf32 matmul.
2026-03-08 18:27:05 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx', 'z', 'x', 'swap'], 'num_qubits': 3, 'num_samples': 100000, 'min_gates': 2, 'max_gates': 12, 'backbone': 'quditkit', 'condition_type': 'UNITARY', 'optimized': True, 'output_path': './tests/datasets/quditkit_unitary_optimized_100k', 'device': 'auto', 'n_jobs': 4}
2026-03-08 18:27:05 - quantum_diffusion.data.dataset - INFO - Generating dataset with 100000 samples, 3 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2026-03-08 18:27:05 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for UNITARY...
[INFO]: Generated 99792 valid circuits.
[INFO]: After filtering unique circuits: 86536.


  0%|          | 0/1 [00:00<?, ?it/s]

 - balance_tensor_dataset, njobs=1, number of samples=86536
 - uniquify_tensor_dataset, number of samples now 86536
 - balancing


  0%|          | 0/26 [00:00<?, ?it/s]

 - dataset size after balancing 86536
[INFO]: allocate memory for U (86536, 2, 8, 8) on cpu approx. 0.041 GB
Split: Train 82210 - Test 4326 

[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_optimized_100k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_optimized_100k/dataset/ds_y.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_optimized_100k/dataset/ds_U.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_optimized_100k/dataset/ds_z.safetensors`.
2026-03-08 18:29:00 - quantum_diffusion.data.dataset - INFO - UNITARY dataset saved to tests/datasets/quditkit_unitary_optimized_100k
2026-03-08 18:29:00 - scripts.generate_dataset - INFO - Dataset generation completed successfully! Took 114.19 seconds.
2026-03-08 18:29:00 - scripts.generate_dataset - INFO - UNITARY: saved to tests/datasets/quditkit_unitary_optimized_100k

DATASET GENERATION SUMMARY
Total time elapsed: 114.19 seconds
Output paths:
  UNIT

### Quditkit without optimization, 10.000 samples

#### Results
- Time: 4.47 seconds
- generated circuits: 9969
- unique circuits: 9340 (93.69%)

In [11]:
cfg.datasets.gate_set = ["h", "cx", "z", "x", "swap"]
cfg.datasets.condition_type = "UNITARY"
cfg.datasets.backbone = "quditkit"
cfg.datasets.optimized = False
cfg.datasets.num_samples = 10000
cfg.datasets.output_path = "./tests/datasets/quditkit_unitary_not_optimized_10k"

main(cfg)

[INFO]: Cuda device has a capability of 8.6 (>= 8), allowing tf32 matmul.
2026-03-08 18:29:00 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx', 'z', 'x', 'swap'], 'num_qubits': 3, 'num_samples': 10000, 'min_gates': 2, 'max_gates': 12, 'backbone': 'quditkit', 'condition_type': 'UNITARY', 'optimized': False, 'output_path': './tests/datasets/quditkit_unitary_not_optimized_10k', 'device': 'auto', 'n_jobs': 4}
2026-03-08 18:29:00 - quantum_diffusion.data.dataset - INFO - Generating dataset with 10000 samples, 3 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2026-03-08 18:29:00 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for UNITARY...
[INFO]: Generated 9969 valid circuits.
[INFO]: After filtering unique circuits: 9340.


  0%|          | 0/1 [00:00<?, ?it/s]

 - balance_tensor_dataset, njobs=1, number of samples=9340
 - uniquify_tensor_dataset, number of samples now 9340
 - balancing


  0%|          | 0/26 [00:00<?, ?it/s]

 - dataset size after balancing 9340
[INFO]: allocate memory for U (9340, 2, 8, 8) on cpu approx. 0.004 GB
Split: Train 8873 - Test 467 

[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_not_optimized_10k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_not_optimized_10k/dataset/ds_y.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_not_optimized_10k/dataset/ds_U.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_not_optimized_10k/dataset/ds_z.safetensors`.
2026-03-08 18:29:04 - quantum_diffusion.data.dataset - INFO - UNITARY dataset saved to tests/datasets/quditkit_unitary_not_optimized_10k
2026-03-08 18:29:04 - scripts.generate_dataset - INFO - Dataset generation completed successfully! Took 4.47 seconds.
2026-03-08 18:29:04 - scripts.generate_dataset - INFO - UNITARY: saved to tests/datasets/quditkit_unitary_not_optimized_10k

DATASET GENERATION SUMMARY
Total time elapsed: 4.47 seconds
Output pat

### Quditkit without optimization, 100.000 samples

#### Results
- Time: 38.54 seconds
- generated circuits: 99787
- unique circuits: 86443 (86.63%)

In [12]:
cfg.datasets.gate_set = ["h", "cx", "z", "x", "swap"]
cfg.datasets.condition_type = "UNITARY"
cfg.datasets.backbone = "quditkit"
cfg.datasets.optimized = False
cfg.datasets.num_samples = 100000
cfg.datasets.output_path = "./tests/datasets/quditkit_unitary_not_optimized_100k"

main(cfg)

[INFO]: Cuda device has a capability of 8.6 (>= 8), allowing tf32 matmul.
2026-03-08 18:29:04 - scripts.generate_dataset - INFO - Dataset configuration: {'gate_set': ['h', 'cx', 'z', 'x', 'swap'], 'num_qubits': 3, 'num_samples': 100000, 'min_gates': 2, 'max_gates': 12, 'backbone': 'quditkit', 'condition_type': 'UNITARY', 'optimized': False, 'output_path': './tests/datasets/quditkit_unitary_not_optimized_100k', 'device': 'auto', 'n_jobs': 4}
2026-03-08 18:29:04 - quantum_diffusion.data.dataset - INFO - Generating dataset with 100000 samples, 3 qubits
[WARNING]: The value 0 is reserved for background tokens, i.e. qubit time position which are not effected by gates.
[WARNING]: Automatically incrementing all vocabulary values by one ...
2026-03-08 18:29:04 - quantum_diffusion.data.dataset - INFO - Starting circuit generation for UNITARY...
[INFO]: Generated 99787 valid circuits.
[INFO]: After filtering unique circuits: 86443.


  0%|          | 0/1 [00:00<?, ?it/s]

 - balance_tensor_dataset, njobs=1, number of samples=86443
 - uniquify_tensor_dataset, number of samples now 86443
 - balancing


  0%|          | 0/26 [00:00<?, ?it/s]

 - dataset size after balancing 86443
[INFO]: allocate memory for U (86443, 2, 8, 8) on cpu approx. 0.041 GB
Split: Train 82121 - Test 4322 

[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_not_optimized_100k/dataset/ds_x.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_not_optimized_100k/dataset/ds_y.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_not_optimized_100k/dataset/ds_U.safetensors`.
[INFO]: Saving tensor to `tests/datasets/quditkit_unitary_not_optimized_100k/dataset/ds_z.safetensors`.
2026-03-08 18:29:43 - quantum_diffusion.data.dataset - INFO - UNITARY dataset saved to tests/datasets/quditkit_unitary_not_optimized_100k
2026-03-08 18:29:43 - scripts.generate_dataset - INFO - Dataset generation completed successfully! Took 38.54 seconds.
2026-03-08 18:29:43 - scripts.generate_dataset - INFO - UNITARY: saved to tests/datasets/quditkit_unitary_not_optimized_100k

DATASET GENERATION SUMMARY
Total time elapsed: 38.54 second